### PR TITLE
Flatpakization

### PR DESCRIPTION
**Pull request by mattiasb**
Make it possible to ship Gnome Twitch as a flatpak.

This pull request is made in haste, and there might be issues with it
but I prefer getting it out there as soon as possible since the changes
touches a lot of code.

The main thing is to use more idiomatic ways of doing things in regards
to the app id, dbus channel, icons, settings and resources etc to make
flatpak-builder happily make packages without complaining.

This was done at GUADEC as part of the flatpak workshop there. A lot
of GNOME people here has been impressed by the quality of gnome-twitch
when I've shown them, so congrats on a great application! :)

### CODE REVIEW

**Comment by vinszent** [comment]
Awesome, thanks for investing your time into this! I was at GUADEC last year (when it was close hehe).

I'm able to build and install it, but I get errors when trying to run it:

    libGL error: No matching fbConfigs or visuals found
    libGL error: failed to load driver: swrast

Which seems to be related to [this](https://github.com/flatpak/flatpak/issues/138).

Seeing as GT is an app related to games, a large amount of people using GT would have either NVIDIA or ATI graphics cards, so this is a blocker.

Personally I've been looking into Snap packages a bit, which seems promising although I've had most success with AppImage in the past. Either way having some kind of universal package is definitely a requirement to be ready for the future.

**Comment by mattiasb** [comment]
> Awesome, thanks for investing your time into this! 
> I was at GUADEC last year (when it was close hehe).

Oh nice! I was part of the organizing team that year. :)

> Seeing as GT is an app related to games, a large amount of people using GT
> would have either NVIDIA or ATI graphics cards, so this is a blocker.

I personally see no harm in shipping a flatpak as a tech preview until this is sorted out. We don't really have an app store in place yet anyways so it's still early. :)

I'll try to revisit these patches soonish and clean them up a bit. Thanks for the enthusiasm! :)

**Comment by vinszent** [comment]
Well I actually got it working :tada: . The new default backend player in v0.3.0 is actually the new GStreamer-Cairo one which doesn't use any OpenGL. I was able to make both a FlatPak package (with some modifications to your script) and a Snap package. I'll integrate your PR as soon as I get the app id sorted out (I think I'll use` com.vinszent.GnomeTwitch` ).

**Comment by mattiasb** [comment]
Wohoo! :)

**Comment by mattiasb** [comment]
>  I think I'll usecom.vinszent.GnomeTwitch

Do you plan on registering that domain btw? Not that the risk of namespace clashes are big, but the namespace are supposed (I think) to reflect a domain you own.

**Comment by vinszent** [comment]
Ya I'm going to register it tonight ;)

**Comment by probonopd** [comment]
>  I've had most success with AppImage in the past

If `ppa:samoilov-lex/gnome-twitch` could provide a deb for trusty, then it would be relatively easy to make an [AppImage](http://appimage.org) out of it that would run on more than just the very latest distributions.

**Comment by vinszent** [comment]
@probonopd Probably won't happen for Trusty because GT depends on a recent version of Meson to build and GTK+ 3.20 which aren't available in 14.04.

**Comment by mattiasb** [comment]
Updated this for latest master.

For testing check out my branch and do:
```
$ git reset HEAD^ --hard
$ flatpak-builder --repo=../repo flatpak com.vinszent.GnomeTwitch.json
```

Before merge do 
```
$ git rebase --interactive --autosquash HEAD^^
```

**Comment by mattiasb** [comment]
Rebased against master again. Any chance this will get merged? :)

**Comment by vinszent** [comment]
Yes of course! I'm trying to get a fileserver up and running (on VPS) to be able to host the OStree. Once I'm done with that I will merge this.

Also I kinda don't want to rename the icon... It seems to be pretty split on whether it should have it or not (looking at GNOME's git). Some icon themes already have variants for the ` gnome-twitch` icon and renaming it would just cause them to break. 

Plus it's technically not included in this list https://wiki.gnome.org/HowDoI/ChooseApplicationID :stuck_out_tongue:

**Comment by mattiasb** [comment]
That's awesome!! Regarding the hosting! :) 

Hm. It seems important for flatpak-builder though, as it won't export it otherwise . I could possibly make the flatpak manifest rename the icon but then we'd still have problems with the icon themes. 

In the old days of "one distro, one set of packages"-world, not namespacing kind of works but since flatpaks are meant to be installed directly from upstream I see conflicts happening rather fast. Hm.

**Comment by mattiasb** [comment]
Regarding gnome git most of them should probably just rename their icons. I know nautilus did just recently as part of becoming a flatpak. :)

**Comment by vinszent** [comment]
Oh well, I guess it's better to do it now rather than later then.

**Comment by mattiasb** [comment]
Yeah, sorry for the trouble. We're walking into new territory here...

File: com.vinszent.GnomeTwitch.json [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "https://github.com/mattiasb/gnome-twitch.git",
+                    "branch": "feature/flatpak"
+                },
+                {
+                    "type": "script",
+                    "dest-filename": "configure",
+                    "commands": [
```
Comment (hadess) [inline com.vinszent.GnomeTwitch.json]: Why not add those as files on the local filesystem instead? I'd create a separate flatpak directory, and put all the files in there.

File: com.vinszent.GnomeTwitch.json [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                    "url": "git://git.gnome.org/libpeas",
+                    "branch": "libpeas-1.18.0"
+                }
+            ]
+        }, {
+            "name": "gnome-twitch",
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "https://github.com/mattiasb/gnome-twitch.git",
```
Comment (hadess) [inline com.vinszent.GnomeTwitch.json]: Wrong URL for merging.

File: com.vinszent.GnomeTwitch.json [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            "cleanup": [ "*" ],
+            "sources": [
+                {
+                    "type": "archive",
+                    "url": "http://www.tortall.net/projects/yasm/releases/yasm-1.3.0.tar.gz",
+                    "sha256": "3dce6601b495f5b3d45b59f7d2492a340ee7e84b5beca17e48f862502bd5603f"
+                }
+            ]
+        }, {
+            "name": "gst-libav",
```
Comment (hadess) [inline com.vinszent.GnomeTwitch.json]: Could be in a separate package as well. See https://github.com/flatpak/flatpak/issues/123

File: com.vinszent.GnomeTwitch.json (hunk 2 above)
Comment (mattiasb) [inline com.vinszent.GnomeTwitch.json]: Yeah I know. I had a !fixup-commit directly after that changed it back. I think I just dropped that for the last version.

File: com.vinszent.GnomeTwitch.json (hunk 1 above)
Comment (mattiasb) [inline com.vinszent.GnomeTwitch.json]: Hm, yeah makes sense.

**Comment by mattiasb** [comment]
Made another PR for an alternate solution based on hadess suggestions #202

**Pull request closed without merging**